src/os/bluestore/Allocator.cc

```diff
@@ -100,6 +107,57 @@ class Allocator::SocketHook : public AdminSocketHook {
       f->open_object_section("fragmentation");
       f->dump_float("fragmentation_rating", alloc->get_fragmentation());
       f->close_section();
+    } else if (command == "bluestore allocator fragmentation histogram " + name) {
+      int64_t num_buckets = 8;
+      cmd_getval(cmdmap, "num_buckets", num_buckets);
+      if (num_buckets < 2) {
+        ss << "Invalid amount of buckets (min=2): '" << num_buckets
+           << std::endl;
+        return -EINVAL;
+      }
+
+      //make 8 buckets to place extent depending on its size
+      // using the following length ranges:
+      // [0..4K] (4K..16K] (16K..64K] .. (4M..16M] (16M..]
+      //
+      std::vector<size_t> counts = {0};
+      counts.resize(num_buckets, 0);
+      size_t base_bits = 12;
+      size_t base = 1ull << base_bits;
+      size_t max = 1ull << (base_bits + 2 * (num_buckets - 2)); //16M
+      auto iterated_allocation = [&](size_t off, size_t len) {
+        int idx;
+        if (len <= base) {
+          idx = 0;
+        } else if (len > max) {
+          idx = num_buckets - 1;
+        } else {
+          size_t most_bit = cbits((uint64_t(len))) - 1;
```

Comment:
NIT: 
most_bit = cbits(len - 1); //len will not be 0, ever
//11..0  -> idx 0
//13..12 -> idx 1
//15..14 -> idx 2
idx = (most_bit - 10) / 2;

Reply:
This required some adjustment since I'd like to include 16K, 64K,.. sizes to previous bucket so we get ranges like that: [0..4K] (4K..16K] (16K..64K] while you formula results in [4K..16K) [16K..64K) ranges..
Anyway thanks for the hint!